[FLINK-27794][connectors/jdbc] Fix the wrong primary key for MysqlCatalog when there are tables with same name in different databases

dusukang opened the pull request:
## What is the purpose of the change
Use MysqlCatalog to get the primary key of the table when there are tables with same name in different databases.

## Brief change log
Pass the database name as catalog parameter for retrieving primary keys by full table identifier in `metaData.getPrimaryKeys(database, schema, table)`

## Verifying this change
This change is a code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

fsk119 reviewed (commented):
Thanks for your contribution. Could you add some basic tests to make sure modification in the future will not break the tests? You can refer to MySqlCatalogITCase.

fsk119 commented:
I don't reproduce the problem in the issue. Could you also tell us which MySql version you use and the mysql driver version?

dusukang commented:
> I don't reproduce the problem in the issue. Could you also tell us which MySql version you use and the mysql driver version?

Thank you very much for your reply, My MySQL version is 8.0.16. There is a table named mysql.`user` in the MySQL, when i create a table named `user` and get the primary key, but the returned primary key is of mysql.`user`.  It while happen when there are tables which have the same name in different databases, when we use catalog to get the primary key for a table, the result may be from other tables which have same names. Can you help me verify it?

This is my case:
[https://issues.apache.org/jira/browse/FLINK-27794](url)

fsk119 commented:
Hi. Thanks for your update. I produce the problem in my local environment after I using the MySql 8.0.16. Let me summarize the problem here: in the MySql 5.7, JDBC uses the scheam as the synonym of the database but in the MySql 8.0 it uses the catalog as the synonym of the database[1].  In the MySql 8.0, it uses the following sql to fetch pks

```
SELECT TABLE_SCHEMA AS TABLE_CAT, NULL AS TABLE_SCHEM, TABLE_NAME, COLUMN_NAME, SEQ_IN_INDEX AS KEY_SEQ, 'PRIMARY' AS PK_NAME FROM INFORMATION_SCHEMA.STATISTICS WHERE TABLE_NAME = 'user' AND INDEX_NAME='PRIMARY' ORDER BY TABLE_SCHEMA, TABLE_NAME, INDEX_NAME, SEQ_IN_INDEX
```

We can notice  `TABLE_CAT` is `NULL` in the SQL. 

So I think the fix is on the correct way. WDYT, @RocMarshal ?

[1] https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-connp-props-connection.html#cj-conn-prop_databaseTerm

RocMarshal commented:
Thanks @dusukang for the driving this and @fsk119 ping. 
+1 to me to fix it.
But I'm not sure whether the current fix is compatible with different versions of MySQL.
Would you mind  giving some test cases for different mysql versions to verify the fix ?

fsk119 commented:
Hi, guys. I push to my repo about how to reproduce the problem[1]. @dusukang  I think you can add a similar test. 

[1] https://github.com/apache/flink/compare/master...fsk119:mysql-catalog-fix?expand=1

dusukang commented:
Thanks @fsk119 for the test cases and @RocMarshal reply. I test different versions of MySQL 5.6.51 and 5.7.35 and 8.0.16 and 8.0.26 locally. It works normally in my test cases, here: 

```
CREATE TABLE flinksql_test01.order_info (
  `order_id` bigint(20) NOT NULL,
  `order_name` varchar(128) DEFAULT NULL,
  `pro_id` bigint(20) DEFAULT NULL,
  `pro_name` varchar(128) DEFAULT NULL,
  PRIMARY KEY (`order_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;

CREATE TABLE flinksql_test02.order_info (
  `order_id` bigint(20) DEFAULT NULL,
  `order_name` varchar(128) DEFAULT NULL,
  `pro_id` bigint(20) NOT NULL,
  `pro_name` varchar(128) DEFAULT NULL,
  PRIMARY KEY (`pro_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

```
import org.apache.flink.connector.jdbc.catalog.MySqlCatalog;
import org.apache.flink.table.catalog.CatalogBaseTable;
import org.apache.flink.table.catalog.ObjectPath;
import org.apache.flink.table.catalog.exceptions.TableNotExistException;

public class MySqlCatalogTest {
    public static void main(String[] args) throws TableNotExistException {
        MySqlCatalog mySqlCatalog =
                new MySqlCatalog(
                        "mysql-catalog",
                        "flinksql_test01",
                        "root",
                        "123456",
                        "jdbc:mysql://127.0.0.1:3307");
        CatalogBaseTable table =
                mySqlCatalog.getTable(new ObjectPath("flinksql_test01", "order_info"));
        System.out.println(table.getUnresolvedSchema());
    }
}
```
@RocMarshal could you have a look?

fsk119 commented:
Hi, all. After digging in, I find the SQL

```
SELECT TABLE_SCHEMA AS TABLE_CAT, NULL AS TABLE_SCHEM, TABLE_NAME, COLUMN_NAME, SEQ_IN_INDEX AS KEY_SEQ, 'PRIMARY' AS PK_NAME FROM INFORMATION_SCHEMA.STATISTICS
WHERE TABLE_NAME = 't_grouped_by_sink' AND INDEX_NAME='PRIMARY' ORDER BY TABLE_SCHEMA, TABLE_NAME, INDEX_NAME, SEQ_IN_INDEX;
```
will get the PK name whose name is `t_grouped_by_sink` only. But the original logic that converts the ResultSet to PK

```
       while (rs.next()) {
            String columnName = rs.getString("COLUMN_NAME");
            pkName = rs.getString("PK_NAME"); // all the PK_NAME should be the same
            int keySeq = rs.getInt("KEY_SEQ");
            System.out.println(columnName); //Print all the column name
            keySeqColumnName.put(keySeq - 1, columnName); // KEY_SEQ is 1-based index
        }
```

may overwrite the pk here. You can notice `keySeqColumnName` is a Map. If you have two tables whose name is equal and they both have pks, it will only keep the PK in the last table in the `keySeqColumnName`.

dusukang commented:
Thanks for your update. I also mentioned this in the previous issue.
 I have debugged this case, it will not happen when we use mysql-connector-java-5.x. in `com.mysql.jdbc.DatabaseMetaData`, it uses the sql to get PK such as:
```
show keys from ${table_name} from ${database_name};
```
but it may happen when we use mysql-connector-java-8.x. in `com.mysql.cj.jdbc.DatabaseMetaDataUsingInfoSchema`, it will use the database value to the catalog when the following condition established, but it's not established, so the catalog is null. then it uses the sql such as @fsk119 provided to get PK
```
        if (catalog == null) {
            if (this.conn.getNullCatalogMeansCurrent()) {
                catalog = this.database;
            }
        }
```

so I think we can do some filtering to resolve tables whose have same name.

fsk119 commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:
```diff
@@ -516,4 +519,8 @@ protected String getSchemaName(ObjectPath tablePath) {
         throw new UnsupportedOperationException();
     }
+
+    protected String getDbCatalog() {
```
It should be like getSchemaName(ObjectPath tablePath)

fsk119 commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalog.java:
```diff
@@ -42,15 +42,16 @@ public class JdbcCatalog extends AbstractJdbcCatalog {
 
     public JdbcCatalog(
             String catalogName,
+            String dbCatalog,
```
We should infer the catalog from the input table path.

fsk119 reviewed (commented):
Thanks for your update. I think we should align with the getSchemaName here rather than introduce an option when building the catalog. The catalog has the abilities to read all metadata in the database instance.

fsk119 commented:
I push my thoughts to the https://github.com/fsk119/flink/pull/new/mysql-catalog

Could you modify the PR as I do and add more tests?

dusukang commented:
Thank you for @fsk119  suggestion. I have modified the PR in this way and added one test. I want to create same tables in two databases to reproduce the case, so should I modify the file `catalog-init-for-test.sql`?

fsk119 commented:
Yes. Just modify as you wish.

fsk119 commented:
Please use rebase and push your code forcibly. In the community, git merge is not suggested.

dusukang commented:
Thanks. I'll try it.

fsk119 commented:
@dusukang Hi. The CI test fails. Could you check whether your modification breaks the test?

dusukang commented:
Yeah. I'll check it.

dusukang commented:
@flinkbot run azure

dusukang commented:
@fsk119 Hi. It has passed the CI test. Could you take a look?

fsk119 commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:
```diff
@@ -144,12 +144,13 @@ public String getBaseUrl() {
         // the returned primary key columns are ordered by COLUMN_NAME, not by KEY_SEQ.
         // We need to sort them based on the KEY_SEQ value.
-        ResultSet rs = metaData.getPrimaryKeys(null, schema, table);
+        ResultSet rs = metaData.getPrimaryKeys(catalog, schema, table);
```
Could you add checks in Line 161? 
```
Preconditions.checkState(!keySeqColumnName.containsKey(keySeq - 1), "The PK constrants should be unique.");

```

fsk119 commented on flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/MySqlCatalogITCase.java:
```diff
@@ -251,6 +267,21 @@ public void testGetTable() throws TableNotExistException {
+                        .column("uid", DataTypes.BIGINT().notNull())
+                        .column("uname", DataTypes.VARCHAR(36))
+                        .column("others", DataTypes.VARCHAR(128))
+                        .primaryKeyNamed("PRIMARY", Lists.newArrayList("uid"))
```
Use Collections.singletonList here.

fsk119 reviewed (commented):
Thanks for your update. It looks good to me in general. I leave some minor comments.

dusukang commented:
Yeah. Thanks for the comment. I also think it is a good improvement to verify. I'll change it soon.

dusukang commented:
@flinkbot run azure

fsk119 reviewed (approved):
Thanks for your update. LGTM.

dusukang commented:
Thanks for your reply.

dusukang commented:
@flinkbot run azure

leonardBang commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:
```diff
@@ -144,19 +144,23 @@ public String getBaseUrl() {
             int keySeq = rs.getInt("KEY_SEQ");
+            Preconditions.checkState(
+                    !keySeqColumnName.containsKey(keySeq - 1),
+                    "The PK constraint should be unique.");
```
Could we improve the exception description? It looks like we meet duplicated records with same primary key.

dusukang commented:
Thanks @leonardBang  for the comments. I agree, I'll change it soon. But If there are other databases you need both `catalog`,`database`,`table` to define a unique table?

leonardBang commented:
> Thanks @leonardBang for the comments. I agree, I'll change it soon. But If there are other databases you need both `catalog`,`database`,`table` to define a unique table?

Currently，Flink Catalog uses three level table path design， we need adjust other databases' table path to this style, otherwise we need a new catalog API design, in this case , a FLIP is required.

dusukang commented:
> > Thanks @leonardBang for the comments. I agree, I'll change it soon. But If there are other databases you need both `catalog`,`database`,`table` to define a unique table?
> 
> Currently，Flink Catalog uses three level table path design， we need adjust other databases' table path to this style, otherwise we need a new catalog API design, in this case , a FLIP is required.

Yeah. Thanks for your reply.

dusukang commented:
@leonardBang Hi. I have modified the PR with your suggestion. Could you have a look?

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@leonardBang Hi. I have updated the pr. After adding the 5.6.x test, it compiled fails. I found that there is no json type in 5.6.x. I made a little change to `catalog-init-for-test.sql`, could you review this when you have time?

leonardBang commented on flink-connectors/flink-connector-jdbc/src/test/resources/mysql-scripts/catalog-init-for-test.sql:
```diff
@@ -59,7 +58,7 @@ CREATE TABLE `t_all_types` (
   `col_integer` int(11) DEFAULT NULL,
   `col_integer_unsigned` int(10) unsigned DEFAULT NULL,
-  `col_json` json DEFAULT NULL,
+  `col_json` longtext DEFAULT NULL,
```
The column name `col_json` and type `longtext` is not matched, we can skip the json type test by removing this field. Although this is a minor test fallback, I think it's okay as we did not cover all types like GEO type too.

leonardBang commented on flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/MySqlCatalogITCase.java:
```diff
@@ -251,6 +267,21 @@ public void testGetTable() throws TableNotExistException {
 
+    @Test
+    public void testGetTablePrimaryKey() throws TableNotExistException {
+        Schema tableSchemaUser =
```
IIUC, we need another table in another database with same name 'user' and different pk 'another_pk'  to validate your PR works

leonardBang commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:
```diff
@@ -144,19 +144,25 @@ public String getBaseUrl() {
             int keySeq = rs.getInt("KEY_SEQ");
+            Preconditions.checkState(
+                    !keySeqColumnName.containsKey(keySeq - 1),
+                    "The PK must be from the same table.");
```
The field(s) of primary key must be from the same table.

leonardBang commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/AbstractJdbcCatalog.java:
```diff
@@ -144,19 +144,25 @@ public String getBaseUrl() {
         // the returned primary key columns are ordered by COLUMN_NAME, not by KEY_SEQ.
         // We need to sort them based on the KEY_SEQ value.
-        ResultSet rs = metaData.getPrimaryKeys(null, schema, table);
+        // In the currently supported databases, database is equivalent to catalog.
```
// In the currently supported database dialects MySQL and Postgres, the database term is equivalent to catalog term. We need to pass the database name as catalog parameter for retrieving primary keys by full table identifier

leonardBang commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalog.java:
```diff
@@ -169,6 +169,11 @@ protected String getSchemaName(ObjectPath tablePath) {
 
+    @Override
+    protected String getDatabaseName(ObjectPath tablePath) {
+        return PostgresTablePath.fromFlinkTableName(tablePath.getObjectName()).getPgSchemaName();
```
Please add a test for this, I remember we use 'catalog.db.\`schema.table\`' in Flink for Postgres table representation, the databaseName is diff with schemaName.

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@leonardBang Hi. Thanks for your comments, I have modified the PR. I found that databaseName can be obtained by `tablePath.getDatabaseName()`, please take a look when you have time.

MartijnVisser commented:
Shouldn't we first get this fixed in `master` before backporting this to `release-1.15` ? @leonardBang  ?

leonardBang commented:
> Shouldn't we first get this fixed in `master` before backporting this to `release-1.15` ? @leonardBang ?

Yes, fixing in master firstly and then backport to other released branches is the right work flow. 
I will have a final review about this PR, please also open a PR for master @dusukang

dusukang commented:
@leonardBang yeah. I will do it soon.

dusukang commented:
I opened a PR for master[https://github.com/apache/flink/pull/20025](https://github.com/apache/flink/pull/20025)

leonardBang commented:
Thanks @dusukang for the update, LGTM. But could you use `git rebase release-1.15` which can avoid conflicts instead of `git merge release-1.5` command ?

dusukang commented:
> Thanks @dusukang for the update, LGTM. But could you use `git rebase release-1.15` which can avoid conflicts instead of `git merge release-1.5` command ?

Yes. I haven't used rebase before... There may be some usage problems. As a result, I have to merge before rebase every time.

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

dusukang commented:
@flinkbot run azure

leonardBang commented:
The failure test is related to https://issues.apache.org/jira/browse/FLINK-28226， let's rebase util the FLINK-28226 fix has been backported to release-1.15, just wait a minute @dusukang . @HuangXingBo is working for the fixing.

dusukang commented:
@leonardBang the ci report build failed on `e2e_n_ci`, it seems that the connectors module is not related. If I am wrong, please help correct it. Is this because of the python3 package version?
```
ERROR: Complete output from command python setup.py egg_info:
Jun 25 15:22:37     ERROR: Traceback (most recent call last):
Jun 25 15:22:37       File "<string>", line 1, in <module>
Jun 25 15:22:37       File "/tmp/pip-install-fixi7jcz/grpcio/setup.py", line 263, in <module>
Jun 25 15:22:37         if check_linker_need_libatomic():
Jun 25 15:22:37       File "/tmp/pip-install-fixi7jcz/grpcio/setup.py", line 213, in check_linker_need_libatomic
Jun 25 15:22:37         stderr=PIPE)
Jun 25 15:22:37       File "/tmp/.conda/lib/python3.7/subprocess.py", line 775, in __init__
Jun 25 15:22:37         restore_signals, start_new_session)
Jun 25 15:22:37       File "/tmp/.conda/lib/python3.7/subprocess.py", line 1522, in _execute_child
Jun 25 15:22:37         raise child_exception_type(errno_num, err_msg, err_filename)
Jun 25 15:22:37     FileNotFoundError: [Errno 2] No such file or directory: 'c++': 'c++'
Jun 25 15:22:37     ----------------------------------------
Jun 25 15:22:37 ERROR: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-fixi7jcz/grpcio/
The command '/bin/sh -c /tmp/.conda/bin/python -m pip install apache-flink-1.15.dev0.tar.gz' returned a non-zero code: 1
```
Sorry to bother you, could you help me see if the upgrade of [https://github.com/apache/flink/pull/20044](https://github.com/apache/flink/pull/20044) is applicable to release-1.15? @MartijnVisser

dusukang commented:
> The failure test is related to [https://issues.apache.org/jira/browse/FLINK-28226，](https://issues.apache.org/jira/browse/FLINK-28226%EF%BC%8C) let's rebase util the [FLINK-28226](https://issues.apache.org/jira/browse/FLINK-28226) fix has been backported to release-1.15, just wait a minute @dusukang . @HuangXingBo is working for the fixing.
Thanks.

dusukang commented:
@flinkbot run azure

dusukang commented:
@leonardBang Hi, It has passed the CI tests. Could you have a look?
